File: javascript/lasagna-master/lasagna-master.js
/// <reference path="./global.d.ts" />
// @ts-check


/**
 * Implement the functions needed to solve the exercise here.
 * Do not forget to export them so they are available for the
 * tests. Here an example of the syntax as reminder:
 * 
 * export function yourFunction(...) {
 *   ...
 * }
 */

/**
 * Check on status of lasagna
 * @param {Number} minutesRemaining 
 * @returns status as string
 */
export function cookingStatus(minutesRemaining)
{
    if (minutesRemaining == undefined || minutesRemaining == null || Number.isNaN(minutesRemaining))
        return 'You forgot to set the timer.'

    if (minutesRemaining <= 0)
        return 'Lasagna is done.'

    return 'Not done, please wait.'
}

/**
 * Calculate lasagna preparation time based on time per layer and given set of layers
 * @param {*} layers list of layers in the lasagna
 * @param {Number} perLayerMinutes how many minutes each layer takes to prepare
 * @returns {Number} preparation time in minutes.
 */
export function preparationTime(layers, perLayerMinutes = 2)
{
    return layers.length * perLayerMinutes
}

/**
 * Calculate the quantities of noodles and sauce required for a given lasagna layer set.
 * @param {*} ingredients a list of ingredients one per layer
 * @returns object with keys noodles and sauce filled with required noodle and sauce in grams and liters respectively
 */
export function quantities(ingredients)
{
    const noodleIncrement = 50;
    const sauceIncrement = 0.2;
    let noodleGrams = 0
    let sauceLiters = 0;
    for(let i = 0; i < ingredients.length; i++)
    {
        if (ingredients[i] == 'noodles')
            noodleGrams += noodleIncrement
        else if (ingredients[i] == 'sauce')
            sauceLiters += sauceIncrement;
    }
    return { noodles: noodleGrams, sauce: sauceLiters}
}


/**
 * Update my list of lasagna to include friends secret ingredient.
 * @param {Array<string>} friendsList friends list of lasagna ingredient last is the secret ingredient.
 * @param {Array<string>} myList my list of lasagna ingredients
 * @returns nothing
 */
export function addSecretIngredient(friendsList, myList)
{
    myList.push(friendsList.slice(-1)[0])
}

/**
 * 
 * @param {Object} recipe object where keys are ingredients and values are the amount of ingredient.
 * @param {Number} servings how many servings we want to make
 * @returns {Object} object with the same recipe but servings scaled up to desired amount.
 */
export function scaleRecipe(recipe, servings)
{
    let ret = new Object()
    let scaleFactor = servings / 2.0

    for(let key in recipe)
    {
        if (recipe.hasOwnProperty(key))
            ret[key] = recipe[key] * scaleFactor
    }

    return ret
}

File: go/lasagna-master/lasagna_master.go
package lasagnamaster

// TODO: define the 'PreparationTime()' function
func PreparationTime(layers []string, timePerLayer int) int {
	if timePerLayer == 0 {
		timePerLayer = 2
	}

	return len(layers) * timePerLayer

}

// TODO: define the 'Quantities()' function
func Quantities(layers []string) (int, float64) {
	noodleLayers := 0
	sauceLayers := 0
	for index := 0; index < len(layers); index++ {
		if layers[index] == "noodles" {
			noodleLayers++
		}
		if layers[index] == "sauce" {
			sauceLayers++
		}
	}
	return 50 * noodleLayers, 0.2 * float64(sauceLayers)
}

// TODO: define the 'AddSecretIngredient()' function
func AddSecretIngredient(friendIngredients, ingredients []string) {
	ingredients[len(ingredients)-1] = friendIngredients[len(friendIngredients)-1]
}

// TODO: define the 'ScaleRecipe()' function
func ScaleRecipe(quantities []float64, numServings int) []float64 {
	var scaled []float64
	scaleFactor := float64(numServings) / 2
	for index := 0; index < len(quantities); index++ {
		scaled = append(scaled, quantities[index]*scaleFactor)
	}
	return scaled
}

// Your first steps could be to read through the tasks, and create
// these functions with their correct parameter lists and return types.
// The function body only needs to contain `panic("")`.
//
// This will make the tests compile, but they will fail.
// You can then implement the function logic one by one and see
// an increasing number of tests passing as you implement more
// functionality.
